Fix code format

Korilakkuma opened the pull request:
### This PR will...

This PR will fix code format by Prettier.

### Why is this Pull Request needed?

This PR is needed to format code that ESLint cannot format.

### Are there any points in the code the reviewer needs to double check?

N/A

### Resolves issues:

N/A

### Checklist

- [x] changes have been done against master branch, and PR does not conflict
- [x] new unit / functional tests have been added (whenever applicable)
- [x] API or design changes are documented in API.md

robwalch commented on .eslintrc.js:
```diff
@@ -35,17 +35,15 @@ module.exports = {
       'unix'
     ],
     // spacing
+    'space-before-function-paren': [2, "never"],
```
I am also not a fan of this space! Thanks.

What do other maintainers/contributors think?

robwalch commented on src/controller/fragment-tracker.ts:
```diff
@@ -17,34 +17,36 @@ export const FragmentState = {
+    | {
+        [key in SourceBufferName]?: TimeRanges;
+      }
+    | null = Object.create(null);
```
I do not like what prettier is doing here. 

Didn't it al add an extra pipe?
```
private timeRanges:
  | {
```

itsjamie replied:
Yeah, that's a typescript thing, you can define the pipe on the first argument to make it easier to re-order the type union later.

robwalch commented on src/controller/timeline-controller.ts:
```diff
@@ -393,76 +398,95 @@ class TimelineController implements ComponentAPI {
+      this.initPTS[frag.cc],
+      vttCCs,
+      frag.cc,
+      cues => {
```
I don't like these kind of line break changes and removal of parens on single param arrow functions.

itsjamie replied:
Yeah, it seems to run counter to the reason for using the multi-line pipes and last element comma support in typescript. Which is not having to add these extra syntax pieces when you change the code slightly.

tjenkinson replied:
Parentheses are back in the next prettier version. We should update!

https://prettier.io/blog/2020/03/21/2.0.0.html#api

robwalch commented on src/controller/timeline-controller.ts:
```diff
@@ -393,76 +398,95 @@ class TimelineController implements ComponentAPI {
+      parseIMSC1(
+        payload,
+        this.initPTS[frag.cc],
+        () => {
```
same. I find this less readable.

itsjamie replied:
In the general-case, I disagree, but where both of these callback functions are so small, yeah, it does make it appear larger in density.

robwalch commented on src/crypt/aes-decryptor.ts:
```diff
@@ -242,10 +242,10 @@ export default class AESDecryptor {
+      t0 = (invSBOX[s0 >>> 24] << 24) ^ (invSBOX[(s1 >> 16) & 0xff] << 16) ^ (invSBOX[(s2 >> 8) & 0xff] << 8) ^ invSBOX[s3 & 0xff] ^ invKeySchedule[ksRow];
+      t1 = (invSBOX[s1 >>> 24] << 24) ^ (invSBOX[(s2 >> 16) & 0xff] << 16) ^ (invSBOX[(s3 >> 8) & 0xff] << 8) ^ invSBOX[s0 & 0xff] ^ invKeySchedule[ksRow + 1];
+      t2 = (invSBOX[s2 >>> 24] << 24) ^ (invSBOX[(s3 >> 16) & 0xff] << 16) ^ (invSBOX[(s0 >> 8) & 0xff] << 8) ^ invSBOX[s1 & 0xff] ^ invKeySchedule[ksRow + 2];
+      t3 = (invSBOX[s3 >>> 24] << 24) ^ (invSBOX[(s0 >> 16) & 0xff] << 16) ^ (invSBOX[(s1 >> 8) & 0xff] << 8) ^ invSBOX[s2 & 0xff] ^ invKeySchedule[ksRow + 3];
```
Are we sure these changes are safe?

Korilakkuma commented on .eslintrc.js:
```diff
@@ -35,17 +35,15 @@ module.exports = {
       'error',
-      2
-    ],
-    'quotes': [
```
I think that formatted code is readable by Prettier in the following case.

- ESLint quotes
  - `'\n''`
- Prettier
  - `"\n"`

itsjamie commented on src/controller/timeline-controller.ts:
```diff
@@ -480,23 +504,23 @@ class TimelineController implements ComponentAPI {
       const tmpByte = byteArray[position++];
-      const ccbyte1 = 0x7F & byteArray[position++];
-      const ccbyte2 = 0x7F & byteArray[position++];
+      const ccbyte1 = 0x7f & byteArray[position++];
```
If there is an option to make this capitalized, we should since that was the perference before.

itsjamie commented:
Overall, I like it. If we can disable the parenthesis grouping rule I'd be ok to merge as-is. If possible, we should add a GitHub action check that runs prettier and commits any changes to open PRs so drive-by PRs don't need to run prettier to get accepted.

itsjamie commented on .prettierrc:
```diff
@@ -0,0 +1,8 @@
+  "tabWidth": 2,
+  "semi": true,
+  "bracketSpacing": true,
+  "printWidth": 180
```
```suggestion
  "singleQuote": true,
  "semi": true,
  "printWidth": 180,
  "trailingComma": "es5",
  "arrowParens": true,
```

itsjamie replied:
I removed the values that matched the defaults. If we're adopting prettier, does anyone mind if we set `"semi": false`? It's going to catch the places we need it when automatic semicolon insertion would fail us?

tjenkinson replied:
not a blocker but personally I like having the semi colons

¯\\_(ツ)_/¯

itsjamie replied:
Will leave it as `semi: false` if anyone cares for them, I don't care enough to remove them :)

itsjamie commented:
I went looking at the prettierrc configuration, and I don't think we can disable the parenthesis grouping changes on the hex code. I assume it will do the bitwise operators correctly, but before merging, someone (maybe me at a later time) should double-check that assumption.

tjenkinson commented:
+1 for prettier.

I wonder if we can also simplify the eslint config and plugins too, and just have `eslint-config-prettier` and `eslint-plugin-prettier` as mentioned on https://www.robertcooper.me/using-eslint-and-prettier-in-a-typescript-project  (also keeping the custom rules we have around restricted globals and properties)?

I also think prettier should run on the pre-commit hook. [There's an example with `pretty-quick here](https://prettier.io/docs/en/precommit.html#option-2-pretty-quickhttpsgithubcomazzpretty-quick)

I think some of the other plugins and rules might not be necessary anymore because they would be impossible to break with prettier anyway, or there's a risk they'll conflict with each other

itsjamie commented on .prettierrc:
```diff
@@ -0,0 +1,7 @@
+  "semi": true,
+  "printWidth": 180,
+  "trailingComma": "es5",
+  "arrowParens": true,
```
```suggestion

```

itsjamie replied:
As @tjenkinson pointed out, this will be changing to default to "always". Meaning we can remove this assuming we go with prettier v2. Which I don't see why we wouldn't.

itsjamie commented on .prettierrc:
```diff
@@ -0,0 +1,7 @@
+  "singleQuote": true,
+  "semi": true,
+  "printWidth": 180,
+  "trailingComma": "es5",
```
```suggestion
  "trailingComma": "es5",
```

itsjamie replied:
This is becoming the default value in Prettier 2.0, we can remove this.

Korilakkuma commented on scripts/precommit.sh:
```diff
@@ -1,5 +1,5 @@
 set -e
 
-npm run lint
+npm run lint:fix
```
This script contains `npm run format` by using `eslint-plugin-prettier`.

tjenkinson commented on .eslintrc.js:
```diff
@@ -118,7 +117,17 @@ module.exports = {
+    'no-prototype-builtins': 0,
+    'prettier/prettier': [
+      2,
+      {
+        'singleQuote': true,
+        'semi': true,
+        'printWidth': 180,
+        'trailingComma': 'es5',
+        'arrowParens': 'always',
+      }
```
is this needed or would it default to reading `.prettierrc` without it?

Korilakkuma replied:
I think that this config is required to combine with ESLint.

https://github.com/prettier/eslint-plugin-prettier#options

tjenkinson commented on .eslintrc.js:
```diff
@@ -18,7 +18,8 @@ module.exports = {
   'extends': [
     'eslint:recommended',
-    'standard'
+    'standard',
```
do we still gain anything from `standard`?

Also please add [`eslint-config-prettier`](https://github.com/prettier/eslint-config-prettier) to prevent any conflicts

Korilakkuma replied:
Because lint errors occur, it seems required ...

tjenkinson commented:
@Korilakkuma I added some commits. Hope you don't mind.

It updates prettier, removes some rules I think were conflicting, and adds a check to makes sure that every file has been run through prettier on CI. Would be worth investigating adding a GitHub action to run prettier automatically as @itsjamie suggested, but I don't think that would need to be in this PR.

robwalch commented:
It would be a good idea to rebase the configuration changes into a single commit so you can drop the prettier changes when the branch changes upstream. feature/v1.0.0 will be updated following releases from master (v0.14.0-x)

Label appropriately if you want to stop the stale bot.

Korilakkuma commented:
@tjenkinson 

Thanks for your commits and comment. Therefore, I dropped some commits for running prettier automatically.

@robwalch 

Thanks. Therefore, I rebased into a single commit.

tjenkinson commented:
@Korilakkuma I think it would be better to remove all the code changes and just keep the config files and package.json etc. We can then do `npm run prettier` just before this is merged to prevent merge conflicts with the actual source code.

I just pushed another commit that does this :)

robwalch commented:
@Korilakkuma @tjenkinson @itsjamie is this still something you want to see in v1.0.0?

If so can we get a rebase or merge to sync with the latest in feature/v1.0.0, and settle on any final changes (like should the precommit run `lint:fix`?) and agree on next steps (gh action?).

Thanks!

tjenkinson commented:
@robwalch I think this would be nice to have. Maybe we could update this against master when https://github.com/video-dev/hls.js/pull/3072/ is merged?

tjenkinson commented on scripts/travis.sh:
```diff
@@ -0,0 +1,94 @@
+#!/bin/bash
```
This file should be deleted now :)

Korilakkuma replied:
Sorry. Now, fixed.

tjenkinson reviewed (commented):
Thanks for bringing it up to date. Also need to add prettier to the dev dependencies. 

Finally does the pre-commit hook commit the changes prettier makes if it does or do they get left behind?

I think the pretty-quick option on https://prettier.io/docs/en/precommit.html#option-2-pretty-quickhttpsgithubcomazzpretty-quick would be good to switch to anyway because it should be faster if we configure it only to run on the changed files.

Korilakkuma commented:
@tjenkinson 

> Finally does the pre-commit hook commit the changes prettier makes if it does or do they get left behind?

The pre-commit hook commits the changes prettier makes. Therefore, this PR adds `pretty-quick` for running prettier on changed files.

tjenkinson commented:
I pushed a few more commits and LGTM!

Sorry I suggested using `pretty-quick` but given we're also using eslint `lint-staged` actually works better.

I also removed the eslint standard plugin because I think this would fight a lot with prettier, and added the eslint prettier plugin to disable any of the built in eslint rules that also confict with prettier. Hopefully no one is attached to keeping the standard plugin?

Finally it will also verify that prettier has run on CI now.

@itsjamie we could probably get a github action to commit fixes but I think with the precommit hook it should happen on the user machine, so I'm not sure if we need it. Also it's easy for anyone to run locally and push if needed. One issue I could see with an action making the commit is that commits with the default github action token will not trigger other actions, which would then cause the PR checks not to run on the new commit I think.

tjenkinson commented:
@robwalch thoughts? IMO this is ready to go

tjenkinson commented:
I just reverted to the default print width, which has made a lot of files longer, but I think this is actually preferable because otherwise you end up having to scroll horizontally to see a full line. Just noticed this with the last commit I added where the array in lint-staged.config.js would always switch to one line which would overflow the screen.

Prettier also has this warning

> For readability we recommend against using more than 80 characters:

> In code styleguides, maximum line length rules are often set to 100 or 120. However, when humans write code, they don’t strive to reach the maximum number of columns on every line. Developers often use whitespace to break up long lines for readability. In practice, the average line length often ends up well below the maximum.

> Prettier’s printWidth option does not work the same way. It is not the hard upper allowed line length limit. It is a way to say to Prettier roughly how long you’d like lines to be. Prettier will make both shorter and longer lines, but generally strive to meet the specified printWidth.

> Remember, computers are dumb. You need to explicitly tell them what to do, while humans can make their own (implicit) judgements, for example on when to break a line.

> In other words, don’t try to use printWidth as if it was ESLint’s max-len – they’re not the same. max-len just says what the maximum allowed line length is, but not what the generally preferred length is – which is what printWidth specifies.

robwalch reviewed (commented):
@tjenkinson I'd like to pull the PR down and make sure everything looks good in my IDE before giving the green light. Give me a few hours to do that this morning.

All the changes look good. Thanks for the attention to detail on this!

tjenkinson commented on README.md:
```diff
@@ -16,43 +16,45 @@ It works by transmuxing MPEG-2 Transport Stream and AAC/MP3 streams into ISO BMF
-[TypeScript]: https://www.typescriptlang.org/
-[TypeScript compiler]: https://www.typescriptlang.org/docs/handbook/compiler-options.html
-[Webpack]: https://webpack.js.org/
+[html5 video]: https://www.html5rocks.com/en/tutorials/video/basics/
+[mediasource extensions]: https://w3c.github.io/media-source/
+[http live streaming]: https://en.wikipedia.org/wiki/HTTP_Live_Streaming
+[web worker]: https://caniuse.com/#search=worker
+[ecmascript6]: https://github.com/ericdouglas/ES6-Learning#articles--tutorials
+[typescript]: https://www.typescriptlang.org/
+[typescript compiler]: https://www.typescriptlang.org/docs/handbook/compiler-options.html
+[webpack]: https://webpack.js.org/
```
Hmm that’s weird not sure why that changed

tjenkinson replied:
Ah prettier is doing this and I think it's because in this format it doesn't matter, because it's a reference to the placeholders further up

tjenkinson replied:
it doesn't change ones in the form `[Link](someurl)`

robwalch reviewed (changes requested):
Can we make prettier ignore api-docs (and any other "dist" dirs)?

I got an error running `npm run sanity-check` twice or after generating docs with `npm run docs`:
```
...
[warn] api-docs/typedef/index.html
[warn] api-docs/variable/index.html
[warn] Code style issues found in the above file(s). Forgot to run Prettier?
```

robwalch commented on .prettierignore:
```diff
@@ -0,0 +1,4 @@
+/dist
```
```suggestion
/dist
/api-docs
```

tjenkinson replied:
Good catch

robwalch commented on src/hls.ts:
```diff
@@ -178,83 +208,125 @@ export default class Hls implements HlsEventEmitter {
-        listener.apply(this, args);
-      } else {
-        try {
+    this._emitter.on(
+      event,
+      function (this: Context, ...args: unknown[]) {
```
I will need to get used to these line breaks. I still prefer seeing event listener registration condensed in a single line.

robwalch replied:
I'll get over it.
